### DISCUSSION TITLE
Connection of websocket server with the generated artifacts

### DISCUSSION THREAD

**Discussion by GSri30** [eclipse-xtext/xtext#2704] [Q&A] [answered]
Hello,

I was exploring the `Xtext` codebase.

*Aim:* To build a language server and integrate it with my monaco-editor in react app (language client is implemented using `monaco-languageclient`) via websockets.

*Current state:* I was looking at few of the existing implementations.
https://github.com/TypeFox/monaco-languageclient/issues/162
https://github.com/nmaoez/composer-language-server

*Doubt:* 
(1) The flow I understood was 
```monaco-editor -> monaco-languageclient -> websockets (client) <-> websockets (server) <- language server <- custom language grammer```
I understand that after writing our custom language grammar, we generate artifacts. Then we implement a websocket server. Then I see we link them by mapping their corresponding `input` and `output` streams. Something like this
https://github.com/nmaoez/composer-language-server/blob/0b65a08b106e7e4a89f4c5be9db7f740ceaf207e/com.novomind.language.server.expression.websockets/src/com/novomind/language/server/expression/websockets/RunWebSocketServer3.xtend#LL44-LL46

But the doubt is, we are just using an injector and injecting a `LanguageServerImpl` class which is common to everyone. How does this know about my grammar and the generated artifacts?
i.e. How is ``` language server <- custom language grammer ``` connection achieved.

**Comment by cdietrich** [comment]
xtext uses Java Service Loader API to detect language servers. therefor the workflow will generate
mydsl.ide/src-gen/META-INF/services/org.eclipse.xtext.ISetup

=> if you put your dsl to classpath. xtext.ide shared language server (started via the launchers calling injector via ServerModule)
will detect your lang automatically if it is on classpath

**Reply by GSri30** [reply]
So, if I understand it well, something like this is happening

```
Launcher<LanguageClient> launcher = Launcher.createLauncher(languageServer,
					LanguageClient.class, args.getIn(), args.getOut(), args.isValidate(), args.getTrace());
```
This maps the socket inp/out to the language servers' inp/out. So when a req comes into the websocket, it is forwarded to the shared language server instance. And internally, using the `Java Service Loader API`, this inp information is communicated with my artifacts..

Am I correct?

Also if this is the case, I wanted to know what should be the general flow one should follow to implement a language server.

- Generate a xtext project in eclipse using the template
- Define the custom grammar
- Generate artifacts
- Create a new file that defines the launchers using the injector via ServerModule (ServerLauncher file) in the `org.xtext.example.mydsl.ide project's src folder ONLY` (Doubt).
- Then rest it handles

**Reply by cdietrich** [reply]
you dont need to create files.
just put the jar created for mydsl + mydsl.ide on the classpath

if you use the new project wizard it will have an option to create a language server app or uber jar that has already everything packaged. (deselect eclipse, select maven or gradle, select under language server)

**Reply by GSri30** [reply]
Yeah, I actually used the new project wizard. I have used the fat-jar config.
So how can I configure it to make it use with my custom websocket port? (Assuming the websockets etc are automatically setup from selecting the option to create lang srv in project wizard, along with the launchers)

**Reply by cdietrich** [reply]
you need to change the config to use your new mainclass.
depends on what you select to do that
by default is uses io main class

**Reply by GSri30** [reply]
Okay. Ig this is the file in action by default.
https://github.com/eclipse/xtext/blob/main/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ServerLauncher.java#LL72-LL73

What could be the required config from my end, assuming I want to establish a websocket connection with the language server.

**Reply by cdietrich** [reply, marked as answer]
you need to check the build.gradle or pom.
there you can specifiy the main class

**Reply by GSri30** [reply]
<img width="746" alt="Screenshot 2023-06-15 at 4 40 26 PM" src="https://github.com/eclipse/xtext/assets/54752266/66cd1c94-8172-4ed5-b4a0-27267c82658c">

So I need to just change the `mainClassName` that's present here.
Any other config required?

**Reply by cdietrich** [reply]
no changing the the main class should be sufficient

**Reply by GSri30** [reply]
Got it.
Thanks for your very quick support!

**Comment by cdietrich** [comment]
![Screenshot 2023-06-15 110454](https://github.com/eclipse/xtext/assets/238736/52bf71c6-8ccd-44ae-b9ca-92c393d54c71)
